pennylane/templates/subroutines/select.py

```diff
@@ -70,44 +80,243 @@ class Select(Operation):
+            0:    ─╭○───────○╮─╭○───────○╮─╭○───────○╮─╭○───────○╮─╭●───────●╮─╭●───────●╮─╭●───────●╮─╭●───────●╮─┤
+            1:    ─├○───────○┤─├○───────○┤─├●───────●┤─├●───────●┤─├○───────○┤─├○───────○┤─├●───────●┤─├●───────●┤─┤
+            aux0:  ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ ╰─╭●───●╮─╯ │
+            2:    ───├○───○┤─────├●───●┤─────├○───○┤─────├●───●┤─────├○───○┤─────├●───●┤─────├○───○┤─────├●───●┤───┤
+            aux1:    ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯     ╰─╭●──╯   │
+            3:    ─────├U(M0)──────├U(M1)──────├U(M2)──────├U(M3)──────├U(M4)──────├U(M5)──────├U(M6)──────├U(M7)──┤
+            4:    ─────╰U(M0)──────╰U(M1)──────╰U(M2)──────╰U(M3)──────╰U(M4)──────╰U(M5)──────╰U(M6)──────╰U(M7)──┤
+
+        Here, we used the symbols
+
+        .. code-block::
+
+            0: ─╭●──       ─●─╮─
+            1: ─├●──  and  ─●─┤─
+            2:  ╰───       ───╯
+
+        for :class:`~.TemporaryAND` and its adjoint, respectively, and skipped drawing the
+        auxiliary wires in areas where they are guaranteed to be in the state :math:`|0\rangle`.
+        We will need three simplification rules for pairs of ``TemporaryAND`` gates:
+
+        .. code-block::
+
+            ─○─╮─╭○──   ──     ─○─╮─╭○──   ─╭○─       ─○─╮─╭●──   ─╭●────
+            ─○─┤─├○── = ──,    ─○─┤─├●── = ─│──, and  ─●─┤─├○── = ─│──╭●─.
+            ───╯ ╰───   ──     ───╯ ╰───   ─╰X─       ───╯ ╰───   ─╰X─╰X─
+
+        Applying these simplifications reduces the computational cost of the ``Select``
+        template:
+
+        .. code-block::
+
+            0:    ─╭○────────────────╭○──────────────────╭●─────────────────────╭●─────────────────●╮─┤
+            1:    ─├○────────────────│───────────────────│──╭●──────────────────│──────────────────●┤─┤
+            aux0:  ╰─╭●─────╭●────●╮─╰X─╭●─────╭●─────●╮─╰X─╰X─╭●─────╭●─────●╮─╰X─╭●─────╭●─────●╮─╯ │
+            2:    ───├○─────│─────●┤────├○─────│──────●┤───────├○─────│──────●┤────├○─────│──────●┤───┤
+            aux1:    ╰─╭●───╰X─╭●──╯    ╰─╭●───╰X──╭●──╯       ╰─╭●───╰X──╭●──╯    ╰─╭●───╰X──╭●──╯   │
+            3:    ─────├U(M0)──├U(M1)─────├U(M2)───├U(M3)────────├U(M4)───├U(M5)─────├U(M6)───├U(M7)──┤
+            4:    ─────╰U(M0)──╰U(M1)─────╰U(M2)───╰U(M3)────────╰U(M4)───╰U(M5)─────╰U(M6)───╰U(M7)──┤
+
+        An additional cost reduction then results from the fact that ``TemporaryAND`` gates
+        (``adjoint(TemporaryAND)`` gates) take four (zero) :class:`~T` gates, instead of the seven
+        ``T`` gates required by a decomposition of :class:`~Toffoli`.
+
+        For general :math:`c` and :math:`K=2^c`, the decomposition takes a similar form, with
+        alternating control and auxiliary wires.
+
+        An implementation of this decomposition is easily achieved in the following steps:
+        We first define a recursive subroutine ``R`` that will be used in the main function:
```

Comment:
```suggestion
        We first define a recursive sub-circuit ``R`` that will be repeatedly applied:
```

I think I was mostly thrown off by the terms "sub-routine" in combination with "main function". What is even meant by the main function here? 😅